### PR TITLE
Don't fail to match when mutable elevated facts change.

### PR DESCRIPTION
**Pull request by roliveri**
# Overview

This PR is being created to address [RHINENG-9012](https://issues.redhat.com/browse/RHINENG-9012).

In the event an elevated canonical fact changes - the insights_id for example - and there are no other elevated facts that match, the current logic prevents matching the host based on secondary canonical facts - like MAC addresses or bios uuid.
This results in duplicate host entries being created in the database.

However, some elevated facts are truly immutable - provider_id for example - if they don't match then the host shouldn't match even if the secondary facts are the same.

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: tests/helpers/test_utils.py [hunk 1]
```diff
@@ -86,17 +86,29 @@ def set_environment(new_env=None):
...
+def base_host(**values):
+    return HostWrapper(_base_host_data(**values))
+
+
+# This method returns a host wrapper that contains a single canonical fact, making it
+# a minimal host that's valid for use with mq_create_or_update_host().
+# It should be used in test cases where specific canonical facts are not needed.
+def minimal_host(**values):
+    host_wrapper = HostWrapper(_base_host_data(**values))
+    host_wrapper.bios_uuid = generate_uuid()
```
Comment (thearifismail) [inline tests/helpers/test_utils.py:110]: What's the reason for choosing `bios_uuid` over `insights_id`?

**Comment by roliveri** [comment]
- @fstavela 

There are two IQE tests failing:

1.  tests/rest/test_deduplication.py::test_dedup_elevated_facts_not_matching[insights_id]
2. tests/rest/test_deduplication.py::test_dedup_elevated_facts_not_matching[subscription_manager_id]

The test expects a new host to be created, however now an existing host is found based on secondary canonical facts.
This new behavior prevents duplicates when the `insights_id` of `subscription_manager_id` changes.

The test uses `create_host_data()` which generates a set of secondary canonical facts. The test only updates the elevated canonical facts leaving the secondary facts unchanged. It should probably delete the secondary facts in addition to deleting the elevated facts that aren't tested field.

NB: the test succeeds for `provider_id/provider_type` because it's considered an immutable canonical fact. When an immutable fact is different, we assume it's a different host even if the secondary facts are the same.

In summary:

- If a mutable elevated canonical fact changes (`insights_id`, `subscription_manager_id`) then match on secondary canonical facts.
- If an immutable elevated canonical fact changes (provider_id/provider_type) then there's no match

File: lib/host_repository.py [hunk 2]
```diff
@@ -84,7 +86,24 @@ def find_existing_host(identity, canonical_facts):
     existing_host = _find_host_by_elevated_ids(identity, canonical_facts)
 
     if not existing_host:
-        existing_host = find_host_by_multiple_canonical_facts(identity, canonical_facts)
+        #
+        # If we didn't find a host based on elevated IDs, it could mean that one or more
+        # of the elevated IDs have changed. We remove the elevated IDs from the following
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/host_repository.py:91]: Here, we're only referring to `insights_id` and `subscription_manager_id`, not `provider_is/provider_type`.

File: lib/host_repository.py [hunk 3]
```diff
@@ -171,6 +190,9 @@ def find_host_by_multiple_canonical_facts(identity, canonical_facts):
     """
     logger.debug("find_host_by_multiple_canonical_facts(%s)", canonical_facts)
 
+    if not canonical_facts:
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/host_repository.py:193]: Some canonical facts are removed from the provided list throughout the process, so there is a chance that the list could be empty after removal.

File: tests/helpers/test_utils.py (hunk 1 above)
Comment (roliveri) [reply tests/helpers/test_utils.py:110]: The purpose of the only canonical fact supplied by this method is just to ensure the request is valid as it stands, so it doesn't matter which fact it chooses. The original version of this method added a fixed IP address, which would result in undesired matches, so it needs to generate a unique fact every call.

The `insights_id` is an elevated canonical fact, so choosing it would require changes in some of the tests that call this function.

File: lib/host_repository.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thearifismail) [reply lib/host_repository.py:91]: It will be very helpful stating explicitly, "If we didn't find a host based on 'insights_id' or 'subscription_manager_id'...."  Why is `provider_id` not included?

File: lib/host_repository.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (thearifismail) [reply lib/host_repository.py:193]: Adding a comment will be helpful to code readers

File: lib/host_repository.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/host_repository.py:91]: Because provider_id is an immutable fact, so if it's different it's definitely not a match.
I think the comments after that make it more clear. However I do think I should change "IDs" to "facts".

**Comment by roliveri** [comment]
@fstavela 
I tried to investigate the test failures, please let me know if I can do anything more to help.

Findings:

A number of the RBAC tests are failing, but I couldn't determine it it's related to these changes.

Of the ci.ext.devshift.net PR build failures, there are 16 failures across 3 files:

tests/rest/test_api_account_access.py::test_delete_multiple_accounts[RHEL]
tests/rest/test_api_account_access.py::test_delete_multiple_accounts[CentOS Linux]
- 102 assert len(prim_hosts.results) == test_hosts
- Not sure, It appears as though only 1 host was created instead of 3.

tests/rest/test_deduplication.py::test_dedup_elevated_facts_priority_not_matching[insights_id]
	161 assert new_host.id != host.id
	Could be because elevated facts have changed, and now we can match even when mutable facts change.

tests/rest/test_deduplication.py::test_dedup_elevated_facts_not_matching[insights_id]
tests/rest/test_deduplication.py::test_dedup_elevated_facts_not_matching[subscription_manager_id]
- 328 assert new_host.id != host.id
- Probably the same as above. Could be because elevated facts have changed, and now we can match even when mutable facts change.

tests/rest/test_deduplication.py::test_dedup_canonical_facts_not_matching[satellite_id]
tests/rest/test_deduplication.py::test_dedup_canonical_facts_not_matching[fqdn]
tests/rest/test_deduplication.py::test_dedup_canonical_facts_not_matching[bios_uuid]
tests/rest/test_deduplication.py::test_dedup_canonical_facts_not_matching[ip_addresses]
- 409 assert updated_host.id != host.id
- Probably the same as above. Could be because elevated facts have changed, and now we can match even when mutable facts change.

tests/rest/test_deduplication.py::test_dedup_of_similar_hosts[insights_id]
tests/rest/test_deduplication.py::test_dedup_of_similar_hosts[subscription_manager_id]
- 547 assert updated_host.id == hosts[1].id
- Probably the same as above. Could be because elevated facts have changed, and now we can match even when mutable facts change.

*** Not sure of the following, as my MASTER copy of the code doesn't seem to correspond to the test logs.

tests/rest/test_pagination.py::test_pagination - iqe_host_inventory_ap...

tests/rest/test_pagination.py::test_unique_results_by_id - assert 1 > 1
- 135 assert total_page > 1

tests/rest/test_pagination.py::test_unique_results_for_id_list - iqe_h...
- 361 response = host_inventory.apis.hosts.get_hosts_by_id_response(...

tests/rest/test_pagination.py::test_unique_results_by_hostname_or_id
- 396 assert total_hosts > num_iterations, "This shouldn't fail unless something big has changed"

tests/rest/test_pagination.py::test_pagination_of_system_profile_fetching

**Review by FabriciaDinizRH — COMMENTED** [review]
Overall, looks good, thanks for all the comments!
I'll wait for the tests to pass and then LGTM it

File: lib/host_repository.py [hunk 4]
```diff
@@ -76,8 +78,10 @@ def find_existing_host(identity, canonical_facts):
     logger.debug("find_existing_host(%s, %s)", identity, canonical_facts)
     existing_host = _find_host_by_elevated_ids(identity, canonical_facts)
 
-    if not existing_host:
-        existing_host = find_host_by_multiple_canonical_facts(identity, canonical_facts)
+    if existing_host:
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/host_repository.py:81]: It's not necessary, but I think it's cleaner as the bulk of the code doesn't need to be indented within the "if" statement.

Sorry, it's the old Ruby developer in me. In Ruby, guard clauses are favored over nested conditionals. In fact, the `rubocop` checker would complain if I did it the other way.

I know this is not Ruby, so I can change it back if people feel strongly about it.

**Review by jpramos123 — COMMENTED** [review]
One comment

File: lib/host_repository.py [hunk 5]
```diff
@@ -91,38 +95,52 @@ def find_existing_host_by_id(identity, host_id):
...
+    elevated_keys = []
+    immutable_facts = {}
     for key in ELEVATED_CANONICAL_FACT_FIELDS:
         if key not in canonical_facts.keys():
             continue
-        elevated_facts[key] = canonical_facts[key]
+
+        if key in IMMUTABLE_CANONICAL_FACTS:
+            immutable_facts[key] = canonical_facts[key]
+            if compound_fact := COMPOUND_CANONICAL_FACTS_MAP.get(key):
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/host_repository.py:106]: The `provider_type` is non unique, but it's required to define the domain in which `provider_id` is unique; so it's considered the secondary component of the compound fact. The compound fact itself is identified by its primary component, `provider_id`. So, when `provider_id` is listed as an `immutable_fact`, its associated `provider_type` is included as well through the `COMPOUND_CANONICAL_FACTS_MAP` dictionary, which maps the primary fact (`provider_id`) with its associated secondary fact (`provider_type`). There's no need to map `provider_type` to `provider_id` because `provider_type` is never used to identify the compound fact.

The `CanonicalFactsSchema` ensures that both `provider_id` and `provider_type` are provided.

**Comment by fstavela** [comment]
Here are some notes from my testing so far:
- The elevated facts priority order is now almost unused, because it is enough when just any 1 elevated fact matches the existing host (regardless of it's priority) to update the host. So, when `provider_id` doesn't exist, we can have a situation where 2 hosts which have the same insights_id (for example after a VM clone), but have different mac addresses and other facts, will show up as a single host in Inventory and they will overwrite their data every time. On the other hand, this probably won't break anything new, because even with the old way the `insights_id` was the highest prio fact, so when it matched it would behave the same way, regardless of other facts. But, it is something to think about and maybe if we kept the old priority logic and just moved `mac_addresses` to be the highest prio fact (after `provider_id`), this problem would be solved. But, then we would have to rely on `mac_addresses` to (almost) never change. I don't know if we want to experiment with it now, or just move forward with your existing PR in it's current state and see how it turns out and maybe do more changes to the logic later - that would probably be a better option.
- `mac_addresses` is the kind of canonical fact which may get obfuscated by customers a lot. Do you know how the obfuscation works? Will the obfuscation of the same mac address always result in the same obfuscated mac address, or is it completely random? If the same mac address obfuscates to a different value every time, then this might be a problem.
- I noticed that the payload will match the existing host if all of the mac addresses in the payload are included in the mac addresses in the existing host. This means that the mac addresses in the payload can be shuffled (in different order), it can be a sublist of the existing host's macs, it can include duplicate mac addresses and it will still match (this is all good). But, if the macs list in the payload is a superlist of the existing host's macs (it includes the same macs + additional ones), then the payload won't match the existing host and a new one will be created. Is this something to worry about and do you think it should match?
- As I pointed out in the previous bullet point, it is enough if the macs in the payload are a sublist of the existing host's macs. What if the payload includes only a single mac address, something like `00:00:00:00:00:00`, or `FF:FF:FF:FF:FF:FF`? It will probably match one of the hosts which include this mac address and update the host's data. Again, is this something to worry about? Should there be a workaround for these addresses?

**Comment by roliveri** [comment]
> Here are some notes from my testing so far:
> 
> * The elevated facts priority order is now almost unused, because it is enough when just any 1 elevated fact matches the existing host (regardless of it's priority) to update the host. So, when `provider_id` doesn't exist, we can have a situation where 2 hosts which have the same insights_id (for example after a VM clone), but have different mac addresses and other facts, will show up as a single host in Inventory and they will overwrite their data every time. On the other hand, this probably won't break anything new, because even with the old way the `insights_id` was the highest prio fact, so when it matched it would behave the same way, regardless of other facts. But, it is something to think about and maybe if we kept the old priority logic and just moved `mac_addresses` to be the highest prio fact (after `provider_id`), this problem would be solved. But, then we would have to rely on `mac_addresses` to (almost) never change. I don't know if we want to experiment with it now, or just move forward with your existing PR in it's current state and see how it turns out and maybe do more changes to the logic later - that would probably be a better option.
> * `mac_addresses` is the kind of canonical fact which may get obfuscated by customers a lot. Do you know how the obfuscation works? Will the obfuscation of the same mac address always result in the same obfuscated mac address, or is it completely random? If the same mac address obfuscates to a different value every time, then this might be a problem.
> * I noticed that the payload will match the existing host if all of the mac addresses in the payload are included in the mac addresses in the existing host. This means that the mac addresses in the payload can be shuffled (in different order), it can be a sublist of the existing host's macs, it can include duplicate mac addresses and it will still match (this is all good). But, if the macs list in the payload is a superlist of the existing host's macs (it includes the same macs + additional ones), then the payload won't match the existing host and a new one will be created. Is this something to worry about and do you think it should match?
> * As I pointed out in the previous bullet point, it is enough if the macs in the payload are a sublist of the existing host's macs. What if the payload includes only a single mac address, something like `00:00:00:00:00:00`, or `FF:FF:FF:FF:FF:FF`? It will probably match one of the hosts which include this mac address and update the host's data. Again, is this something to worry about? Should there be a workaround for these addresses?

Thanks for the thoughtful comments. Please see below:

1. As for you first bullet point, given `mac_adresses` is the second highest priority after `provider_id`, when there's no `provider_id` hosts with the same `insights_id` should be identified by `mac_adresses` first, and not be considered the same host. Hence they should not show up as a single host.
2. In discussions I've had about obfuscation, I was lead to believe that the facts that are most likely to be obfuscated are `fqdn` and `ip_addresses` - no mention was made about `mac_addresses`. Having said that, I also made it clear that all obfuscation should be done in a consistent manner, using the same one way hashing algorithm - however I don't believe that is the case today. Bottom line is, if any or all facts can be obfuscated or redacted, there's very little we can do.
3. Yes, you're correct about the `mac_addresses` comparison. That's because we use a jsonb operator to do the list comparison for us. This could be a problem when a new network interface is added to an existing host, but then, we'll probably match on lower-priority facts instead. So, for example, if a network interface was added AND the `insights_id` changed at the same time, we could get a duplicate. However, I think this is far more unlikely than the current situation, where we could get a duplicate whenever the `insights_id` changes.
4. As for your final bullet point. We have to assume consistent collection of all `mac_addresses` available on the host from all reporters. For the host to exist on the network it would have to have at least one unique mac address (other than `00:00:00:00:00:00`, or `FF:FF:FF:FF:FF:FF`).

**Comment by fstavela** [comment]
> > Here are some notes from my testing so far:
> > 
> > * The elevated facts priority order is now almost unused, because it is enough when just any 1 elevated fact matches the existing host (regardless of it's priority) to update the host. So, when `provider_id` doesn't exist, we can have a situation where 2 hosts which have the same insights_id (for example after a VM clone), but have different mac addresses and other facts, will show up as a single host in Inventory and they will overwrite their data every time. On the other hand, this probably won't break anything new, because even with the old way the `insights_id` was the highest prio fact, so when it matched it would behave the same way, regardless of other facts. But, it is something to think about and maybe if we kept the old priority logic and just moved `mac_addresses` to be the highest prio fact (after `provider_id`), this problem would be solved. But, then we would have to rely on `mac_addresses` to (almost) never change. I don't know if we want to experiment with it now, or just move forward with your existing PR in it's current state and see how it turns out and maybe do more changes to the logic later - that would probably be a better option.
> > * `mac_addresses` is the kind of canonical fact which may get obfuscated by customers a lot. Do you know how the obfuscation works? Will the obfuscation of the same mac address always result in the same obfuscated mac address, or is it completely random? If the same mac address obfuscates to a different value every time, then this might be a problem.
> > * I noticed that the payload will match the existing host if all of the mac addresses in the payload are included in the mac addresses in the existing host. This means that the mac addresses in the payload can be shuffled (in different order), it can be a sublist of the existing host's macs, it can include duplicate mac addresses and it will still match (this is all good). But, if the macs list in the payload is a superlist of the existing host's macs (it includes the same macs + additional ones), then the payload won't match the existing host and a new one will be created. Is this something to worry about and do you think it should match?
> > * As I pointed out in the previous bullet point, it is enough if the macs in the payload are a sublist of the existing host's macs. What if the payload includes only a single mac address, something like `00:00:00:00:00:00`, or `FF:FF:FF:FF:FF:FF`? It will probably match one of the hosts which include this mac address and update the host's data. Again, is this something to worry about? Should there be a workaround for these addresses?
> 
> Thanks for the thoughtful comments. Please see below:
> 
>     1. As for you first bullet point, given `mac_adresses` is the second highest priority after `provider_id`, when there's no `provider_id` hosts with the same `insights_id` should be identified by `mac_adresses` first, and not be considered the same host. Hence they should not show up as a single host.
> 
>     2. In discussions I've had about obfuscation, I was lead to believe that the facts that are most likely to be obfuscated are `fqdn` and `ip_addresses` - no mention was made about `mac_addresses`. Having said that, I also made it clear that all obfuscation should be done in a consistent manner, using the same one way hashing algorithm - however I don't believe that is the case today. Bottom line is, if any or all facts can be obfuscated or redacted, there's very little we can do.
> 
>     3. Yes, you're correct about the `mac_addresses` comparison. That's because we use a jsonb operator to do the list comparison for us. This could be a problem when a new network interface is added to an existing host, but then, we'll probably match on lower-priority facts instead. So, for example, if a network interface was added AND the `insights_id` changed at the same time, we could get a duplicate. However, I think this is far more unlikely than the current situation, where we could get a duplicate whenever the `insights_id` changes.
> 
>     4. As for your final bullet point. We have to assume consistent collection of all `mac_addresses` available on the host from all reporters. For the host to exist on the network it would have to have at least one unique mac address (other than `00:00:00:00:00:00`, or `FF:FF:FF:FF:FF:FF`).

Thank you for the responses.

1. It would only be identified by `mac_addresses` if the host with the new `mac_addresses` already exists in the Inventory. But, imagine this scenario: I have a system registered in the Inventory. I will create a clone of this system - it will have the same `insights_id`, but different `mac_addresses`. When I try to register this new cloned system, it will match on the `insights_id` even though it has different `mac_addresses`, because we no longer use the priority in this case and we allow for the higher priority facts to be updated. So, the payload will match the existing host and update the `mac_addresses`. Next time the original system sends payload, it will again match on the `insights_id` and update `mac_addresses`. This will go on until one of these 2 systems updates its `insights_id`. As I said, I don't think this is a blocker, because even though we didn't allow the higher priority facts to change in the past, the `insights_id` had the highest priority, so in this particular scenario nothing really changes. But, I think this is something to note and we can try to solve this cloning issue later.
2. I agree
3. I agree that this is so unlikely, that it is not worth spending our time on it right now. I created a Jira with minor priority for this in case we don't have anything more important to do (that will probably never happen), but also to have something to link to in the tests with a reason for why we are not testing the "super-list" scenario: https://issues.redhat.com/browse/RHINENG-11591
4. I agree with the assumption that the system needs to have a valid mac address, but we would also have to assume that all of the collectors and reporters also correctly provide all of the system's mac addresses. I think we should check in the DB (or in the logs from the incoming kafka messages) that we really never receive a payload with only one of the mentioned mac addresses.

**Comment by roliveri** [comment]
@fstavela #1 is a very good point.
We can probably minimize the possibility of that occurring by flagging `mac_addresses` as an immutable  fact (even though it's really not). Then we may only have an issue when `mac_addresses` are added or replaced. I'll try making that change to see how it works out. We should probably hold off on merging this until I do that.

As for #4, the new version of canonical facts schema requires `mac_addresses`, however it can't validate that they are reported consistently between reporters - but it does state that they should be.

**Comment by fstavela** [comment]
> @fstavela #1 is a very good point. We can probably minimize the possibility of that occurring by flagging `mac_addresses` as an immutable fact (even though it's really not). Then we may only have an issue when `mac_addresses` are added or replaced. I'll try making that change to see how it works out. We should probably hold off on merging this until I do that.
> 
> As for #4, the new version of canonical facts schema requires `mac_addresses`, however it can't validate that they are reported consistently between reporters - but it does state that they should be.

@roliveri Good suggestion on the first point, I agree that making mac_addresses immutable would help in this scenario, but if we do that, we have to be 100% sure that the MAC addresses are reported consistently across reporters, otherwise we will end up with more duplicates. Here is a followup Jira ticket for this: [RHINENG-11657](https://issues.redhat.com/browse/RHINENG-11657)

Unfortunately (and this is related also to point number 4), I found out that the MAC addresses are not reported consistently. I ran a gabi query on Prod and I found out that there are 140 hosts with a single MAC address equal to `00:00:00:00:00:00`. 140 hosts is a very small number considering this is from Prod, but these hosts are from various orgs, including non-internal customer orgs. This means that we are receiving some payloads where `mac_addresses == [00:00:00:00:00:00]`, which can be potentially dangerous with this PR. A created a Jira ticket to investigate why this happens and make sure it's fixed: [RHINENG-11658](https://issues.redhat.com/browse/RHINENG-11658)
When I found this out, I created [this test case](https://gitlab.cee.redhat.com/insights-qe/iqe-host-inventory-plugin/-/merge_requests/1152/diffs?commit_id=4272f49252e0d43d2fe02784a854adb6fe3c6bce):
I have 2 hosts. They don't have provider_id/provider_type. Each of them has 2 MAC addresses, one of them is the same (for example `00:00:00:00:00:00`) and the other one is different. All other canonical and elevated facts are different on each host. Then, I send a payload with only a single MAC address (the one that is the same on both hosts - `00:00:00:00:00:00`) and other canonical and elevated facts from one of the existing hosts. It doesn't matter which one's canonical facts I use, the payload will always update the more recently updated host. I think this should be changed and in case the mac address (or any of the higher prio facts, but this will probably mostly happen in this scenario with mac addresses) match multiple hosts, we should go on to lower prio facts and pick one of these two hosts based on the other facts. @roliveri What do you think?

**Review by fstavela — APPROVED** [review]
LGTM, thank you for making these changes!

**Review by jpramos123 — APPROVED** [review]

**Review by kruai — APPROVED** [review]

**Review by thearifismail — APPROVED** [review]
